File: py/src/jsonrpc3/handler.py
"""Request handler for JSON-RPC 3.0."""

from typing import Any, List, Union

from .errors import (
    RpcError,
    internal_error,
    invalid_reference_error,
    reference_not_found_error,
    reference_type_error,
)
from .params import NULL_PARAMS, new_params
from .protocol import ProtocolHandler
from .session import RpcObject, Session
from .types import (
    Request,
    Response,
    VERSION_30,
    is_notification,
    is_reference,
    new_error_response,
    new_response,
)


class Handler:
    """Handles JSON-RPC 3.0 requests and routes to appropriate objects."""

    def __init__(
        self,
        session: Session,
        root_object: RpcObject,
        caller: Any,
        mime_types: List[str] = None,
    ):
        """
        Initialize handler.

        Args:
            session: Session for reference management
            root_object: Root RPC object for method dispatch
            caller: Caller for making callbacks
            mime_types: Supported MIME types
        """
        self.session = session
        self.root_object = root_object
        self.caller = caller
        self.protocol = ProtocolHandler(session, mime_types)

    def handle_request(self, req: Union[dict, Request]) -> Response:
        """
        Handle a single request.

        Args:
            req: Request message (dict or Request object)

        Returns:
            Response message
        """
        try:
            # Determine which object to call
            ref = req.ref if isinstance(req, Request) else req.get("ref")

            if ref == "$rpc":
                obj = self.protocol
            elif ref:
                local_obj = self.session.get_local_ref(ref)
                if not local_obj:
                    raise reference_not_found_error(ref)
                obj = local_obj
            else:
                obj = self.root_object

            # Extract parameters
            params_value = req.params if isinstance(req, Request) else req.get("params")
            params = new_params(params_value) if params_value is not None else NULL_PARAMS

            # Call method
            method = req.method if isinstance(req, Request) else req["method"]
            result = obj.call_method(method, params, self.caller)

            # Process result for automatic reference registration
            processed_result = self._process_result(result)

            # Create response
            jsonrpc = req.jsonrpc if isinstance(req, Request) else req.get("jsonrpc", VERSION_30)
            req_id = req.id if isinstance(req, Request) else req.get("id")
            return new_response(processed_result, req_id, jsonrpc)

        except RpcError as e:
            # Create error response
            jsonrpc = req.jsonrpc if isinstance(req, Request) else req.get("jsonrpc", VERSION_30)
            req_id = req.id if isinstance(req, Request) else req.get("id")
            return new_error_response(e.to_dict(), req_id, jsonrpc)

        except Exception as e:
            # Convert to internal error
            rpc_error = internal_error(str(e))
            jsonrpc = req.jsonrpc if isinstance(req, Request) else req.get("jsonrpc", VERSION_30)
            req_id = req.id if isinstance(req, Request) else req.get("id")
            return new_error_response(rpc_error.to_dict(), req_id, jsonrpc)

    def handle_batch(self, batch: List[Union[dict, Request]]) -> List[Response]:
        """
        Handle a batch request.

        Args:
            batch: List of requests

        Returns:
            List of responses (excluding notifications)
        """
        responses = []
        batch_results = []

        for i, req in enumerate(batch):
            try:
                # Resolve batch-local references
                resolved_ref = req.ref if isinstance(req, Request) else req.get("ref")
                if resolved_ref and resolved_ref.startswith("\\"):
                    resolved_ref = self._resolve_batch_local_ref(
                        resolved_ref, i, batch_results
                    )

                # Create modified request with resolved ref
                if isinstance(req, Request):
                    resolved_req = Request(
                        jsonrpc=req.jsonrpc,
                        method=req.method,
                        params=req.params,
                        id=req.id,
                        ref=resolved_ref,
                    )
                else:
                    resolved_req = {**req, "ref": resolved_ref}

                # Handle the request
                resp = self.handle_request(resolved_req)

                # Store result for batch-local reference resolution
                batch_results.append(resp.result if isinstance(resp, Response) else resp.get("result"))

                # Only add response if not a notification
                is_notif = is_notification(req)
                if not is_notif:
                    responses.append(resp)

            except Exception as e:
                # Store None for failed requests
                batch_results.append(None)

                # Only add error response if not a notification
                is_notif = is_notification(req)
                if not is_notif:
                    rpc_error = e if isinstance(e, RpcError) else internal_error(str(e))
                    jsonrpc = req.jsonrpc if isinstance(req, Request) else req.get("jsonrpc", VERSION_30)
                    req_id = req.id if isinstance(req, Request) else req.get("id")
                    responses.append(new_error_response(rpc_error.to_dict(), req_id, jsonrpc))

        return responses

    def _resolve_batch_local_ref(
        self, ref: str, current_index: int, results: List[Any]
    ) -> str:
        """Resolve batch-local reference (\\0, \\1, etc.)."""
        try:
            # Parse index from \\N format
            index = int(ref[1:])

            # Check for forward reference
            if index >= current_index:
                raise invalid_reference_error(
                    ref, f"Forward reference not allowed: {ref} at index {current_index}"
                )

            # Check bounds
            if index < 0 or index >= len(results):
                raise invalid_reference_error(ref, f"Batch-local reference out of bounds: {ref}")

            # Get the result
            result = results[index]
            if result is None:
                raise invalid_reference_error(ref, f"Referenced request failed: index {index}")

            # Result must be a Reference
            if not is_reference(result):
                raise reference_type_error(
                    ref, f"Result at index {index} is not a reference"
                )

            return result["$ref"]

        except ValueError:
            raise invalid_reference_error(ref, f"Invalid batch-local reference format: {ref}")

    def _process_result(self, result: Any) -> Any:
        """Process result to handle automatic reference registration."""
        # Nil or basic types - return as-is
        if result is None or isinstance(result, (str, int, float, bool)):
            return result

        # If it's already a Reference, return as-is
        if is_reference(result):
            return result

        # Check if result implements RpcObject
        if self._is_rpc_object(result):
            # Register and return Reference
            ref = self.session.add_local_ref(result)
            return {"$ref": ref}
        elif isinstance(result, list):
            # Process array elements
            return [self._process_result(item) for item in result]
        elif isinstance(result, dict):
            # Process dict values
            return {k: self._process_result(v) for k, v in result.items()}
        else:
            # Other types return as-is
            return result

    def _is_rpc_object(self, value: Any) -> bool:
        """Check if value implements RpcObject interface."""
        return isinstance(value, RpcObject) or (
            hasattr(value, "call_method") and callable(value.call_method)
        )

File: ruby/lib/jsonrpc3/handler.rb
# frozen_string_literal: true

module JSONRPC3
  # Handler routes requests to objects
  class Handler
    attr_reader :session, :root_object

    def initialize(session, root_object, caller, mime_types = [MIME_TYPE_JSON])
      @session = session
      @root_object = root_object
      @caller = caller
      @protocol = ProtocolHandler.new(session, mime_types)
    end

    # Handle a single request
    def handle_request(req)
      # Determine which object to call
      ref = req.is_a?(Request) ? req.ref : req["ref"]

      obj = if ref == "$rpc"
              @protocol
            elsif ref
              local_obj = @session.get_local_ref(ref)
              raise JSONRPC3.reference_not_found_error(ref) unless local_obj

              local_obj
            else
              @root_object
            end

      # Extract parameters
      params_value = req.is_a?(Request) ? req.params : req["params"]
      params = params_value ? JSONRPC3.new_params(params_value) : NULL_PARAMS

      # Call method
      method = req.is_a?(Request) ? req.method : req["method"]
      result = obj.call_method(method, params, @caller)

      # Process result for automatic reference registration
      processed_result = process_result(result)

      # Create response
      jsonrpc = req.is_a?(Request) ? req.jsonrpc : (req["jsonrpc"] || VERSION_30)
      req_id = req.is_a?(Request) ? req.id : req["id"]
      JSONRPC3.new_response(processed_result, req_id, jsonrpc)
    rescue RpcError => e
      # Create error response
      jsonrpc = req.is_a?(Request) ? req.jsonrpc : (req["jsonrpc"] || VERSION_30)
      req_id = req.is_a?(Request) ? req.id : req["id"]
      JSONRPC3.new_error_response(e.to_h, req_id, jsonrpc)
    rescue StandardError => e
      # Convert to internal error
      rpc_error = JSONRPC3.internal_error(e.message)
      jsonrpc = req.is_a?(Request) ? req.jsonrpc : (req["jsonrpc"] || VERSION_30)
      req_id = req.is_a?(Request) ? req.id : req["id"]
      JSONRPC3.new_error_response(rpc_error.to_h, req_id, jsonrpc)
    end

    # Handle a batch request
    def handle_batch(batch)
      responses = []
      batch_results = []

      batch.each_with_index do |req, i|
        begin
          # Resolve batch-local references
          resolved_ref = req.is_a?(Request) ? req.ref : req["ref"]
          if resolved_ref&.start_with?("\\")
            resolved_ref = resolve_batch_local_ref(resolved_ref, i, batch_results)
          end

          # Create modified request with resolved ref
          resolved_req = if req.is_a?(Request)
                           Request.new(**req.to_h.merge(ref: resolved_ref))
                         else
                           req.merge("ref" => resolved_ref)
                         end

          # Handle the request
          resp = handle_request(resolved_req)

          # Store result for batch-local reference resolution
          batch_results[i] = resp.is_a?(Response) ? resp.result : resp["result"]

          # Only add response if not a notification
          is_notif = req.is_a?(Request) ? JSONRPC3.notification?(req) : !req.key?("id")
          responses << resp unless is_notif
        rescue StandardError => e
          # Store nil for failed requests
          batch_results[i] = nil

          # Only add error response if not a notification
          is_notif = req.is_a?(Request) ? JSONRPC3.notification?(req) : !req.key?("id")
          unless is_notif
            rpc_error = e.is_a?(RpcError) ? e : JSONRPC3.internal_error(e.message)
            jsonrpc = req.is_a?(Request) ? req.jsonrpc : (req["jsonrpc"] || VERSION_30)
            req_id = req.is_a?(Request) ? req.id : req["id"]
            responses << JSONRPC3.new_error_response(rpc_error.to_h, req_id, jsonrpc)
          end
        end
      end

      responses
    end

    private

    # Resolve batch-local reference (\0, \1, etc.)
    def resolve_batch_local_ref(ref, current_index, results)
      # Parse index from \N format
      index_str = ref[1..]
      index = Integer(index_str, 10)

      # Check for forward reference
      if index >= current_index
        raise JSONRPC3.invalid_reference_error(ref, "Forward reference not allowed: #{ref} at index #{current_index}")
      end

      # Check bounds
      if index.negative? || index >= results.size
        raise JSONRPC3.invalid_reference_error(ref, "Batch-local reference out of bounds: #{ref}")
      end

      # Get the result
      result = results[index]
      raise JSONRPC3.invalid_reference_error(ref, "Referenced request failed: index #{index}") if result.nil?

      # Result must be a Reference
      unless JSONRPC3.reference?(result)
        raise JSONRPC3.reference_type_error(ref, "Result at index #{index} is not a reference")
      end

      result["$ref"]
    rescue ArgumentError
      raise JSONRPC3.invalid_reference_error(ref, "Invalid batch-local reference format: #{ref}")
    end

    # Process result to handle automatic reference registration
    def process_result(result)
      # Nil or basic types - return as-is
      return result if result.nil? || result.is_a?(String) || result.is_a?(Numeric) ||
                       result.is_a?(TrueClass) || result.is_a?(FalseClass)

      # If it's already a Reference, return as-is
      return result if JSONRPC3.reference?(result)

      # Check if result implements RpcObject (duck typing)
      if rpc_object?(result)
        # Register and return Reference
        ref = @session.add_local_ref(result)
        { "$ref" => ref }
      elsif result.is_a?(Array)
        # Process array elements
        result.map { |item| process_result(item) }
      elsif result.is_a?(Hash)
        # Process hash values
        result.transform_values { |value| process_result(value) }
      else
        # Other types return as-is
        result
      end
    end

    # Check if value implements RpcObject interface
    def rpc_object?(value)
      value.respond_to?(:call_method)
    end
  end
end
